executable-spec/src/Cardano/Ledger/Spec/STS/Update/Hupdate.hs

```diff
@@ -78,11 +98,59 @@ instance  STS (HUPDATE hashAlgo) where
           -- Add newly revealed (but stable) SIPs to the active sips. Note that
           -- we place these new sips as arguments of the left hand side of the
           -- 'Map.union', since this operation is left biased.
-          asips' = (wrsips ▷<= (slot -. (2 *. k))) `Map.union` asips
+          asips' = ( Map.mapWithKey  -- update asips slot with voting period end slot
+                       (\sph _ -> slot `addSlot` (votPeriodEnd sph))
+                       (wrsips ▷<= (slot -. (2 *. k)))
+
+                   )
+                   `Map.union`
+                   asips
+
+          votPeriodEnd siphash =  Data.vpDurationToSlotCnt
```

Comment:
We have some duplication here. Maybe we can define a function:

```haskell
votePeriodEnd :: Hash SIP -> Map (Data.SIPHash hashAlgo) (Data.SIP hashAlgo) -> SlotCount
```

Reply:
Done OK